refactor: color transfer and mnist

TarnMaVas opened the pull request.

TarnMaVas committed:
feature: color transport dashboard visualization

TarnMaVas committed:
refactor: added modules necessary for visualization to pixi

TarnMaVas committed:
refactor: color transfer tests

TarnMaVas committed:
refactor: mnist minor fixes

TarnMaVas committed:
refactor: mnist config update

TarnMaVas committed:
refactor: added documentation for color transfer

TarnMaVas committed:
refactor: added color transfer docs to index.md

TarnMaVas committed:
feature: imporoved color transfer - better problem sampling, rng control, enabled measurement function choice and column drop

aipyth commented on configs/color_transfer/example.yaml:
```diff
@@ -11,6 +11,16 @@ solvers:
+output-dir: ./outputs/color_transfer
+rng-seed: 42
+
+drop-columns:
```
Move this feature into another PR as it would be better to have that for runners (synthetic exp) as well -- just a general feature.

TarnMaVas replied:
I didn't move it as it depends on previous commits and changing that seemed overly messy, but corresponding MNIST changes are already in a new PR

aipyth commented on README.md:
```diff
@@ -92,6 +92,40 @@ $ pixi run benchmark --config configs/runners/gaussians.yaml --folds 1 --export
+batch-size: 100000
+pair-number: 2
+images-dir: ./datasets/images
+output-dir: ./outputs/color_transfer
```
pls update docs to reflect the current state of code

aipyth commented on configs/color_transfer/example.yaml:
```diff
@@ -11,6 +11,16 @@ solvers:
+
+experiment: 
+  name: Time and GPU usage test
+  function: uot.experiments.measurement.measure_time_and_output
```
is it possible to use other measurement functions?

TarnMaVas replied:
Yes, but the function has to return the calculated transport path among its metrics, since it is needed for the transfer. Not all measurement functions do that currently as far as I can see. I will add this note to the docs, didn't get to updating them yet

aipyth commented on pixi.toml:
```diff
@@ -22,3 +23,17 @@ lint = "ruff check ."
+
+[pypi-dependencies]
+jax = { version = ">=0.4.38, <0.7", extras = ["cpu"] }
+gpu-tracker = "*"
```
1. fix eof issue (this "stop" sign that is shown)
2. currently, there is a conflict of pixi files with main branch, which prohibits merging

aipyth commented on tests/test_color_transfer.py:
```diff
@@ -0,0 +1,191 @@
+    transport_image,
+    _transform_numpy,
+)
+import uot.experiments.real_data.color_transfer.color_transfer_metrics as ct_metrics
```
which test util do you use? in repo all other tests are written with pytest

TarnMaVas replied:
pytest, I didn't include the import since I didn't use any explicit features from the module, like mark.parametrize, but it still works via pytest

aipyth commented on uot/experiments/real_data/color_transfer/color_transfer.py:
```diff
@@ -85,6 +86,22 @@ def get_image_problems(data: dict[str, ImageData], image_pairs: list[tuple]) ->
+
+    min_images = 2
+    while min_images * (min_images - 1) < pair_num * 2:
+        min_images += 1
```
what is this while cycle?

aipyth commented on uot/experiments/real_data/color_transfer/color_transfer.py:
```diff
@@ -85,6 +86,22 @@ def get_image_problems(data: dict[str, ImageData], image_pairs: list[tuple]) ->
+    while min_images * (min_images - 1) < pair_num * 2:
+        min_images += 1
+    
+    sampled_images = rng.choice(all_images, size= min(min_images, len(all_images)), replace=False)
```
yes, we would like not to see a pair with repeated image, BUT why do we need to need to remove any repetitions across all pairs?

aipyth commented on uot/experiments/real_data/color_transfer/color_transfer.py:
```diff
@@ -85,6 +86,22 @@ def get_image_problems(data: dict[str, ImageData], image_pairs: list[tuple]) ->
+    sampled_images = rng.choice(all_images, size= min(min_images, len(all_images)), replace=False)
+
+    data = {name: ImageData(name, bin_num) for name in sampled_images}
+    image_pairs = list(itertools.permutations(data, 2))[:pair_num]
```
you still use permutations... guys, there are simpler and more efficient procedures, for example this one:
```
We require: exactly `k` pairs
Input: list of keys (images' filenames), number k of pairs
1. result := []
2. while result arrays does not contain k pairs:
3.     sample integers `i` and `j` to be distinct: these represent the elements of keys array
4.     pair := (keys[i], keys[j])
5.     drop pair if it is already in result
6.     otherwise add pair into result array
7. return result
```

TarnMaVas replied:
Sorry, it seems I misunderstood the issue - I thought that the main problem with the previous approach was that we make a lot of redundant permutations and load redundant images, so I decided to first calculate the number of images required to get this much distinct pairs and then only work with them

aipyth commented on uot/experiments/real_data/mnist_classification/mnist_classification.py:
```diff
@@ -111,14 +106,15 @@ def calculate_results(X: ArrayLike, y: ArrayLike, distance: ArrayLike, indices:
 
     X, y, _ = load_mnist_data()
-    X, y = X[:250], y[:250]
+    X, y = X[:250], y[:250]  # Limit to 250 samples for faster computation
```
can be set in the configuration.

it is just a good tone not to leave any magic numbers in code. if possible -- set as a constant and put that in the beginning of the file.

TarnMaVas replied:
Sorry, I was convinced that I deleted this one but it seems that I forgot, my bad

aipyth commented on uot/experiments/real_data/mnist_classification/mnist_classification.py:
```diff
@@ -111,14 +106,15 @@ def calculate_results(X: ArrayLike, y: ArrayLike, distance: ArrayLike, indices:
         config = yaml.safe_load(file) 
 
     solver_configs = load_solvers(config=config)
+    costs_dir = config.get('costs-dir', 'results/costs')
```
what if someone does not specify `costs-dir` and then wonders where the costs are written?
what if `results/costs` contains some data that is already important and then you write your own there?

be careful with paths. if you're the only one using this software that is perfectly fine, okay.
but it would be better to drop a warning or error into stderr that costs-dir is not specified

aipyth commented on uot/experiments/real_data/mnist_classification/mnist_classification.py:
```diff
@@ -145,8 +141,7 @@ def calculate_results(X: ArrayLike, y: ArrayLike, distance: ArrayLike, indices:
 
-    script_dir = os.path.dirname(os.path.abspath(__file__))
-    export_folder = os.path.join(script_dir, "results")
+    export_folder = config.get('output-dir', 'results')
```
same comment as for `costs-dir`

TarnMaVas committed:
fix: color transfer sampling with no permutations

TarnMaVas committed:
fix: mnist magic numbers removed, missing dir behavior

TarnMaVas committed:
refactor: mnist tests

TarnMaVas committed:
Update README.md

TarnMaVas committed:
refactor: updated docs

TarnMaVas committed:
refactor: final docs tidy-up

aipyth committed:
refactor(color-transfer): reorganize module and enhance logging

aipyth committed:
refactor(color-transfer): force float64

aipyth committed:
feat(color-transfer): implement ColorTransferProblem and ColorTransferExperiment classes; update image processing and metrics measurement

aipyth committed:
refactor(color-transfer): streamline image transformation functions

aipyth committed:
fix(measurement): update _wait_jax_finish to use jax.tree.map instead of tree_map

aipyth committed:
refactor(measurement): update image processing to use JAX instead of NumPy

aipyth committed:
fix(measurement): disable JAX usage in distribution metrics and ensure integer type for indices in pixel transformation

aipyth committed:
fix(metrics): ensure src_support and tgt_support are converted to numpy arrays in compute_wasserstein_distance

aipyth committed:
fix(color-transfer): added use_cost_matrix option in experiment

aipyth committed:
fix(bfm): remove rho_mu_push from metrics

aipyth committed:
chore: cleanup old configs and junk files

aipyth committed:
config: add synthetic dataset yaml configs

aipyth committed:
config: add many runners configs

aipyth committed:
fix(measure): update to_discrete method to filter weights based on include_zeros flag

aipyth committed:
feat(experiment): add learning rate finder and refactor measurement functions

aipyth committed:
fix(generator): adjust mean and variance coefficients for Gaussian Mixture Generator

aipyth committed:
fix(pdlp): update marginal handling to use discrete measures directly

aipyth committed:
Implement AMSGradSolver for entropic optimal transport with multi-marginal support

- Added AMSGradSolver class for 2-marginal dual gradient ascent using AMSGrad optimization.
- Introduced JIT-compiled core loop for efficient computation of transport plans and costs.
- Refactored gradient ascent logic to support multiple marginals (N=1..3) in GradientAscentMultiMarginalSGD.
- Enhanced error handling for cost shape compatibility with marginals.
- Removed deprecated LogDomainGradientAscentSolver and associated functions.
- Updated SAGA solver to ensure compatibility with discrete measures without including zeros.

aipyth committed:
fix(back-and-forth: implement Monge map functionality and refactor gradient computation

- Added `monge_map.py` with functions to compute Monge maps from potential functions in ND.
- Introduced `monge_field_from_psi_nd` to compute the Monge field based on a given potential.
- Implemented `interp_vector_field_at_positions_nd` for multilinear interpolation of vector fields.
- Refactored the gradient computation by moving `_central_gradient_nd` to `uot.utils.central_gradient_nd`.
- Updated `pushforward.py` to utilize the new gradient function and adjusted the computation of `s_raw`.
- Modified `solver.py` to compute the Monge map using the new `monge_map_from_psi_nd` function.
- Cleaned up unused code and comments in `solver.py` related to diagnostics and metrics.

aipyth committed:
feat(sinkhorn): add sinkhorn_divergence_with_solver function and normalize cost in SinkhornTwoMarginalLogJaxSolver

aipyth committed:
feat(metrics): add various utility functions for transport metrics and computations

aipyth committed:
chore(playground): added single py files of solver implementation testing

aipyth committed:
chore(notebooks): added examples of benchmarking in notebooks

aipyth committed:
fix(benchmark): update timestamp format in export filename to include full date

aipyth committed:
feat(scripts): add run_color_transfer script for color transfer benchmarking

aipyth committed:
feat(color-transfer): implement sinkhorn divergence computation and update measurement metrics

aipyth committed:
chore(.gitignore): add data folders and GPU tracker files to ignore list

aipyth committed:
feat(back-and-forth): refactor and enhance the back-and-forth solver and related utilities

- Updated import paths for `_central_gradient_nd` and `cic_pushforward_nd` to improve clarity and organization.
- Implemented a new central gradient function in `pushforward.py` to handle boundary conditions more effectively.
- Enhanced `adaptive_pushforward_nd` to support adaptive sub-cell sampling and improved vertex-based pushforward calculations.
- Added utility functions for computing axis spacings and cell volumes in `generate_nd_grid.py`.
- Improved the `BackNForthSqEuclideanSolver` to allow for flexible pushforward function resolution and added extra metrics for diagnostics.
- Updated the convexity metric function to use modern type hints.
- Minor adjustments to ensure consistent handling of grid shapes and types across various functions.

aipyth committed:
fix(experiment): update learning rate selection method and refactor pushforward imports

aipyth committed:
chore(notebooks): added sinkhorn sparsity tests

aipyth committed:
feat(generators): enhance weight preparation and grid generation in various generator classes

aipyth committed:
feat(generators): update dataset count and borders for grid generator configurations

aipyth committed:
feat(config): update parameter grids for gradient schedules in adam.yaml

aipyth committed:
fix(notebooks): updated some back-and-forth plots

aipyth committed:
Merge branch 'main' into refactor/color_transfer_mnist
